### ISSUE TITLE
Edited photo path missing on Monterey

### ISSUE THREAD

**Issue by RhetTbull** [RhetTbull/osxphotos#871]
Sorry for the delay in providing answers to your questions. Here they go...
> 
> Could you try this:
> 
> Find photo `P2071699` in your library. Verify that Photos shows the file as edited and try to export it using Photos (do not use "export unmodified original", use "export 1 photo"). What I'm trying to figure out is if Photos itself knows where the edited image is or if its just showing you the preview image.
* The "export 1 photo" worked correctly.
> 
> Another test would be to try to edit the edited photo in Photos. Does Photo show the edit screen or do you get an error about missing photo?
* Hitting "Edit" it brings up the Photos Editing screen.

> If Photos can find it and osxphotos can't, I'll need some additional data to figure out the problem.

* I'll upgrade from 0.54.4 to 0.55.1 and give it a go.
* for now the `osxphotos inspect` of `0.55.1` continues not to show the Edite pic Path:
```
Path original: /Volumes/Macintosh HD/Users/Shared/Pictures/Pictures/iPhoto/2009-03-10-2126-29/P2071699.JPG
Path preview: /Users/Shared/Pictures/iPhoto Shared Library.photoslibrary/resources/derivatives/9/9D3E0182-87C9-4A58-ADCC-765C8512178B_1_202_a.jpeg
```

_Originally posted by @oPromessa in https://github.com/RhetTbull/osxphotos/discussions/856#discussioncomment-4395084_

**Comment by RhetTbull** [comment]
@oPromessa when you have time, please upgrade to v0.55.3 and run the following script (copy the python code and save it to `edited_paths.py`)

`osxphotos run edited_paths.py > debug.txt 2>debug_err.txt`

then send me the debug files at rturnbull+git@gmail.com or post them here.  Will contain full path to your library and filename of photos but no other sensitive info.

`edited_paths.py`:

```python
""" Test edited paths for issue #871

    Upgrade at least osxphotos version 0.55.3 (https://github.com/RhetTbull/osxphotos/releases/tag/v0.55.3)

    Save as "edited_paths.py" then run with command:

    osxphotos run edited_paths.py > debug.txt 2>debug_err.txt 

    Then post debug.txt and debug_err.txt here or send to me directly at rturnbull+git@gmail.com

    If you need to specify the path to the library pass it as the first argument to the script:

    osxphotos run edited_paths.py /path/to/library > debug.txt 2>debug_err.txt
"""

import csv
import os
import sys

import osxphotos


def verbose(msg):
    print(msg, file=sys.stderr)
    sys.stderr.flush()


def get_renders(photo):
    """Return list of render resources for given photo"""
    library = photo._db._library_path
    directory = photo._uuid[0]  # first char of uuid
    resource_dir = os.path.join(library, "resources", "renders", directory)
    resources = []
    if os.path.exists(resource_dir):
        resources.extend(
            filename for filename in os.listdir(resource_dir) if photo._uuid in filename
        )
    return resources


def main():
    verbose(f"osxphotos version: {osxphotos._version.__version__}")
    dbfile = sys.argv[1] if len(sys.argv) > 1 else None
    photosdb = osxphotos.PhotosDB(dbfile=dbfile, verbose=verbose)
    path_data = []
    for photo in photosdb.photos():
        if not photo.hasadjustments:
            continue

        candidate_path = photo._path_5()
        candidate_path_edited = photo._path_edited_5()
        renders = get_renders(photo)
        path_data.append(
            {
                "uuid": photo.uuid,
                "filename": photo.original_filename,
                "has_adjustments": photo.hasadjustments,
                "got_path": photo.path is not None,
                "path": photo.path,
                "path_candidate": candidate_path,
                "got_path_edited": photo.path_edited is not None,
                "path_edited": photo.path_edited,
                "path_edited_candidate": candidate_path_edited,
                "renders": ",".join(renders) if renders else "None",
            }
        )

    # print results as CSV
    fieldnames = path_data[0].keys()
    writer = csv.DictWriter(sys.stdout, fieldnames=fieldnames)
    writer.writeheader()
    for row in path_data:
        writer.writerow(row)


if __name__ == "__main__":
    osxphotos.debug.set_debug(True)
    main()
```

**Comment by oPromessa** [comment]
Will attach the files shortly... 

For one example I looked at... he was looking for file:
```
/Users/Shared/Pictures/iPhoto Shared Library.photoslibrary/resources/renders/A/ABDA9CB7-99C0-44FC-9097-3B048CAF88D0_1_201_a.jpeg
```

But the file does not exist on such folder. What does exist is a plist file with a similar file name `ABDA9CB7-99C0-44FC-9097-3B048CAF88D0` without `_1_201_a`...
```
/Users/Shared/Pictures/iPhoto Shared Library.photoslibrary/resources/renders/A/ABDA9CB7-99C0-44FC-9097-3B048CAF88D0.plist 
```

and the contents of this file seems to indicate there is an adjustment with some data. Wild guess here... does Photos only marks the file has been rotated Anti-clockwise and avoids creating the edited jpg file to __save space__?

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
        <key>adjustmentBaseVersion</key>
        <integer>0</integer>
        <key>adjustmentData</key>
        <data>
        PU7LCoJAFP2Xsx5kfCAyP1BtCgpqES2uOuaEozJzdSP+ez6i3eG8J1jNVBIT1ARLnrU7
        avOuGSqO0kT8uIcpuYaKkiQT6JzRLRObroXKZoGqc5b4rp3fqFBg3PGprbq1OB9MU54H
        m2sHhSg8xJGEAPX9P4QklYEMwk3wRa0tXfVodlEK9A3xurM4LRWXG5ZdKj+DZ7uc8VDP
        1/wF
        </data>
        <key>adjustmentEditorBundleID</key>
        <string>com.apple.Photos</string>
        <key>adjustmentFormatIdentifier</key>
        <string>com.apple.photo</string>
        <key>adjustmentFormatVersion</key>
        <string>1.4</string>
        <key>adjustmentTimestamp</key>
        <date>2022-12-18T22:42:20Z</date>
</dict>
</plist>
```

**Comment by RhetTbull** [comment]
>  Wild guess here... does Photos only marks the file has been rotated Anti-clockwise and avoids creating the edited jpg file to save space?

For RAW files, I have seen this happen (specifically on Big Sur). See #225.  However, on Ventura, edits to RAW appear to produce the edited render. But I've not done extensive testing. 

It's possible there are certain edits for which Photos does not produce a corresponding edited image and simply renders at run time from the plist.  I've not seen this in my testing though with exception of aforementioned RAW images.

**Comment by RhetTbull** [comment]
Scanning the debug data, it appears that a lot (maybe all) of the missing edited images does have the adjustments plist file. Take a look at a couple of these and see if you can determine a pattern in either type of image or type of edit that was done on the image.

**Comment by oPromessa** [comment]
Going thru it. Do you now what's the encoding of the "data" file on .plist file. It seems they are a few sets of cases that repeat across the files.

**Comment by RhetTbull** [comment]
Some of them can be decode, available as `PhotoInfo.adjustments`. (Thanks to @neilpa who did some great reverse engineering on this)

Try this:

`osxphotos repl`

That opens an interactive run/eval/print/loop (REPL) environment with your Photos data loaded by osxphotos. Then select an image in Photos and type:

`photo = get_selected()[0]; print(photo.adjustments.adjustments if photo.adjustments else None)`

If the photo has adjustments, it may print out something like this (this example for a RAW photo converted to Black & White):

```python
[
    {
        'formatVersion': 1,
        'enabled': True,
        'settings': {'offsetTone': 0, 'offsetGrain': 0, 'offsetStrength': 0, 'offsetNeutralGamma': 0, 'auto': False, 'inputBlackAndWhite': -0.32291666666666663},
        'identifier': 'SmartBlackAndWhite',
        'formatIdentifier': 'com.apple.photo'
    },
    {
        'formatVersion': 1,
        'enabled': True,
        'settings': {'boostVersion': 'FB3', 'boostParams': [0.42899999022483826, 0.8050000071525574, 0.9573039412498474], 'inputMethodVersion': '8'},
        'identifier': 'RKRawDecodeOperation',
        'formatIdentifier': 'com.apple.photo'
    }
]
```

You can then hit up-arrow then return to execute the same command after selecting another photo. Sometimes the adjustment data cannot be decoded.

**Comment by oPromessa** [comment]
For one pic in particular, yields no adjustments!!!
```python
>>> photo = get_selected()[0]; print(photo.adjustments.adjustments if photo.adjustments else None)
[]
```

The photo.adjustments field points out to the .plist file.
```python
>>> print(photo.adjustments)
AdjustmentsInfo(plist_file='/Users/Shared/Pictures/iPhoto Shared Library.photoslibrary/resources/renders/A/ABDA9CB7-99C0-44FC-9097-3B048CAF88D0.plist')
```

But it does reference an orientation change... `orientation=8` while `original_orientation=1`
```python
>>> print(photo)
uuid: ABDA9CB7-99C0-44FC-9097-3B048CAF88D0
filename: PC310133-a-2.JPG
original_filename: PC310133-a-2.JPG
(...)
ismissing: false
hasadjustments: true
external_edit: false
favorite: false
hidden: false
latitude: null
longitude: null
path_edited: null
shared: false
isphoto: true
ismovie: false
uti: public.jpeg
burst: false
live_photo: false
path_live_photo: null
iscloudasset: false
incloud: null
date_modified: '2022-12-18T22:42:20.473628+00:00'
portrait: false
screenshot: false
slow_mo: false
time_lapse: false
hdr: false
selfie: false
panorama: false
has_raw: false
uti_raw: null
path_raw: null
place: null
exif: ExifInfo(flash_fired=False, iso=80, metering_mode=5, sample_rate=None, track_format=None,
  white_balance=0, aperture=3.0, bit_rate=None, duration=None, exposure_bias=0.0,
  focal_length=10.309999999999999, fps=None, latitude=None, longitude=None, shutter_speed=0.025,
  camera_make='OLYMPUS IMAGING CORP.  ', camera_model='uD800,S800      ', codec=None,
  lens_model=None)
(...)
intrash: false
height: 3264
width: 2448
orientation: 8
original_height: 2448
original_width: 3264
original_orientation: 1
original_filesize: 1770975
```

**Comment by RhetTbull** [comment]
Here's an updated script to collect data on the adjustments.  Please run this the same way and send the data. (This one won't output any paths).

`osxphotos run edited_paths2.py > debug2.txt 2>debug2_err.txt `

```python
""" Test edited paths for issue #871

    Upgrade at least osxphotos version 0.55.3 (https://github.com/RhetTbull/osxphotos/releases/tag/v0.55.3)

    Save as "edited_paths2.py" then run with command:

    osxphotos run edited_paths2.py > debug2.txt 2>debug2_err.txt 

    Then post debug.txt and debug_err.txt here or send to me directly at rturnbull+git@gmail.com

    If you need to specify the path to the library pass it as the first argument to the script:

    osxphotos run edited_paths2.py /path/to/library > debug2.txt 2>debug2_err.txt
"""

import csv
import os
import sys

import osxphotos


def verbose(msg):
    print(msg, file=sys.stderr)
    sys.stderr.flush()


def get_renders(photo):
    """Return list of render resources for given photo"""
    library = photo._db._library_path
    directory = photo._uuid[0]  # first char of uuid
    resource_dir = os.path.join(library, "resources", "renders", directory)
    resources = []
    if os.path.exists(resource_dir):
        resources.extend(
            filename for filename in os.listdir(resource_dir) if photo._uuid in filename
        )
    return resources


def main():
    verbose(f"osxphotos version: {osxphotos._version.__version__}")
    dbfile = sys.argv[1] if len(sys.argv) > 1 else None
    photosdb = osxphotos.PhotosDB(dbfile=dbfile, verbose=verbose)
    path_data = []
    for photo in photosdb.photos():
        if not photo.hasadjustments:
            continue

        candidate_path = photo._path_5()
        candidate_path_edited = photo._path_edited_5()
        renders = get_renders(photo)
        path_data.append(
            {
                "uuid": photo.uuid,
                "filename": photo.original_filename,
                "has_adjustments": photo.hasadjustments,
                "got_path": photo.path is not None,
                "got_path_edited": photo.path_edited is not None,
                "orientation": photo.orientation,
                "original_orientation": photo.original_orientation,
                "orientation_change": photo.orientation != photo.original_orientation,
                "adjustments": bool(photo.adjustments),
                "adjustment_data": bool(photo.adjustments.adjustments)
                if photo.adjustments
                else False,
            }
        )

    # print results as CSV
    fieldnames = path_data[0].keys()
    writer = csv.DictWriter(sys.stdout, fieldnames=fieldnames)
    writer.writeheader()
    for row in path_data:
        writer.writerow(row)


if __name__ == "__main__":
    osxphotos.debug.set_debug(True)
    main()
```

**Comment by oPromessa** [comment]
Thanks for the support!

Please find attached
* [debug2.txt](https://github.com/RhetTbull/osxphotos/files/10287289/debug2.txt)
* [debug2_err.txt](https://github.com/RhetTbull/osxphotos/files/10287292/debug2_err.txt)

**Comment by RhetTbull** [comment]
Thanks -- will take a look. Quick examination shows 884 images missing an edited path.  Of these, 862 appear to be only an orientation change (photo was rotated) as the photo shows as edited but there's no adjustment data in the plist file. Interestingly there appear to be 2143 images where there is an edited image path that also appear to be only an orientation change.

For the images where the edited path is missing and the only edit was an orientation change, osxphotos could conceivably do the rotation upon export (either through Quartz or through exiftool).  For the others, I'm not sure what can be done if the files are truly not in the library.  One idea might be to use PhotoKit to request the adjusted image in which case Photos would perform the adjustments in the plist and return the edited image.  I've had reports of PhotoKit causing issues on Monterey (#625) so can't be 100% sure this will work until we test it.

Can you try the following and let me know if it works?

`osxphotos export /path/to/export --skip-original-if-edited --download-missing --use-photokit --uuid 1DA18E86-596B-4610-BE90-AAE89FAE4522 --verbose`

If that works, then we may have a workable solution.

**Comment by oPromessa** [comment]
Bummer !

```python
21:31:41 MSP edited_paths[2129]> mkdir export ; osxphotos export ./export --skip-original-if-edited --download-missing --use-photokit --uuid 1DA18E86-596B-4610-BE90-AAE89FAE4522 --verbose
osxphotos version 0.55.3
Using last opened Photos library: /Users/Shared/Pictures/iPhoto Shared Library.photoslibrary
Created export database /Users/Shared/Pictures/Export/edited_paths/export/.osxphotos_export.db
Processing database /Users/Shared/Pictures/iPhoto Shared Library.photoslibrary/database/photos.db
Processing database /Users/Shared/Pictures/iPhoto Shared Library.photoslibrary/database/Photos.sqlite
Processing database.
Database version: 6000, 7.
Processing persons in photos.
Processing detected faces in photos.
Processing albums.
Processing keywords.
Processing photo details.
Processing import sessions.
Processing additional photo details.
Processing face details.
Processing photo labels.
Processing EXIF details.
Processing computed aesthetic scores.
Processing comments and likes for shared photos.
Processing moments.
Done processing details from Photos library.
Exporting 1 photo to /Users/Shared/Pictures/Export/edited_paths/export...
Exporting P5130204.JPG (P5130204.JPG) (1/1)
Skipping original version of P5130204.JPG
Exporting edited version of P5130204.JPG (P5130204.JPG)
Skipping missing edited photo P5130204.JPG (1DA18E86-596B-4610-BE90-AAE89FAE4522)
Exporting 1 photos ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00
Processed: 1 photo, exported: 0, missing: 1, error: 0
Elapsed time: 0:00:00
```

**Comment by RhetTbull** [comment]
Back to the drawing board. I'll see
I'll take a look at which methods to instrument via --watch to get additional debug info.

**Comment by RhetTbull** [comment]
Is this still an issue? I now have an old machine running Monterey for testing but was not able to replicate this.

**Comment by oPromessa** [comment]
Sorry for the late reply! Still open this issue 😢 

* I'm not using iCloud but I noticed that using `--download-missing` option allows `osxphotos` to "find" the "missing edited file". I'm using `--skip-original-if-edited`.

* PS: I've rotated many of the original pics and did an export/import (aka transform) cycle and the impact of this issue was largely reduced. I have currently ~40 pics and ~10 Videos in this status.
   * BTW, Trimmed movies in Photos should be exported with the `--skip-original-if-edited`, correct?

### without --download-missing

```py
$ osxphotos export X   --name PC290013 --load-config ~/Pictures/Pictures/Apps/colibri.toml 
2024-07-27 23:50:30.007471 -- Loaded options from file /Users/MSP/Pictures/Pictures/Apps/colibri.toml
(...)
)2024-07-27 23:50:51.437386 -- Done processing details from Photos library.
2024-07-27 23:50:54.166530 -- Exporting 1 photo to /Users/MSP/Documents/GitHub/Export/recreate/working/X...
2024-07-27 23:50:54.193393 -- Edited file for PC290013.JPG is missing, exporting original
(...)
)2024-07-27 23:50:54.964789 -- Exported new file /Users/MSP/Documents/GitHub/Export/recreate/XXX/PC290013.JPG
2024-07-27 23:50:54.965664 -- Processed: 1 photo, exported: 1, updated: 0, skipped: 0, updated EXIF data: 1, missing: 0, error: 0, touched date: 0
2024-07-27 23:50:54.966229 -- Elapsed time: 0:00:00
(...)
````

### with --download-missing

```py
$ osxphotos export X --download-missing   --name PC290013 --load-config ~/Pictures/Pictures/Apps/colibri.toml 
2024-07-27 23:51:42.332292 -- Loaded options from file /Users/MSP/Pictures/Pictures/Apps/colibri.toml
(...)
2024-07-27 23:52:03.374716 -- Done processing details from Photos library.
2024-07-27 23:52:06.114033 -- Exporting 1 photo to /Users/MSP/Documents/GitHub/Export/recreate/working/X...
2024-07-27 23:52:06.165215 -- Exporting PC290013.JPG (PC290013.JPG) (1/1)
2024-07-27 23:52:06.165655 -- Skipping original version of PC290013.JPG
2024-07-27 23:52:06.166856 -- Exporting edited version of PC290013.JPG (PC290013.JPG)
2024-07-27 23:52:07.875761 -- Writing metadata with exiftool for PC290013_edited.jpeg
2024-07-27 23:52:08.830492 -- Exported PC290013.JPG to /Users/MSP/Documents/GitHub/Export/recreate/XXX/PC290013_edited.jpeg
2024-07-27 23:52:08.844068 -- Exported new file /Users/MSP/Documents/GitHub/Export/recreate/XXX/PC290013_edited.jpeg
2024-07-27 23:52:08.844612 -- Touched date on file /Users/MSP/Documents/GitHub/Export/recreate/XXX/PC290013_edited.jpeg
2024-07-27 23:52:08.845185 -- Processed: 1 photo, exported: 1, updated: 0, skipped: 0, updated EXIF data: 1, missing: 0, error: 0, touched date: 1
```

**Comment by RhetTbull** [comment]
Is this an optimized library or "download originals" library?

**Comment by RhetTbull** [comment]
> BTW, Trimmed movies in Photos should be exported with the --skip-original-if-edited, correct?

Yes. I tested it and the trimmed video is exported in this case.

**Comment by oPromessa** [comment]
> Is this an optimized library or "download originals" library?

I'm on Photos "Version 7.0 (460.0.120)" on "Monterey (12.7.5)" and I have:
* iCloud Photos option disabled: following options are greyed out and disabled:
   * Download Originals
   * Optimise Mac Storage

**Comment by RhetTbull** [comment]
> I'm not using iCloud but I noticed that using --download-missing option allows osxphotos to "find" the "missing edited file". I'm using --skip-original-if-edited. Is this expected?

@oPromessa I think I know what's going on but I'm not sure what the best solution is. Based on the data you've provided and your particular configuration, I think that certain edits you've performed on Monterey have resulted in no rendered edited image being generated. The observed behavior on all versions of MacOS I've tried is that an edit generates the rendered image but this isn't happening in your case.

For example, on my Ventura test library, rotating an image resulted in the following new files being generated in the Photos library:

```
   /Users/rhet/Pictures/Test-13.0.0.photoslibrary/resources/renders/3
   /Users/rhet/Pictures/Test-13.0.0.photoslibrary/resources/renders/3/3DD2C897-F19E-4CA6-8C22-B027D5A71907.plist
   /Users/rhet/Pictures/Test-13.0.0.photoslibrary/resources/renders/3/3DD2C897-F19E-4CA6-8C22-B027D5A71907_1_201_a.jpeg
   /Users/rhet/Pictures/Test-13.0.0.photoslibrary/private/com.apple.photolibraryd/caches/clientservertransaction/FB10A0E0-05A8-427E-8921-A1B0531AE25B
```

and the following files being modified (note the thumbnail images are modified):

```
    /Users/rhet/Pictures/Test-13.0.0.photoslibrary/database/Photos.sqlite-wal
    /Users/rhet/Pictures/Test-13.0.0.photoslibrary/database/search/psi.sqlite-shm
    /Users/rhet/Pictures/Test-13.0.0.photoslibrary/database/search/psi.sqlite-wal
    /Users/rhet/Pictures/Test-13.0.0.photoslibrary/resources/renders
    /Users/rhet/Pictures/Test-13.0.0.photoslibrary/resources/journals
    /Users/rhet/Pictures/Test-13.0.0.photoslibrary/resources/derivatives/3
    /Users/rhet/Pictures/Test-13.0.0.photoslibrary/resources/derivatives/thumbs/3356.ithmb
    /Users/rhet/Pictures/Test-13.0.0.photoslibrary/resources/derivatives/thumbs/4133.ithmb
    /Users/rhet/Pictures/Test-13.0.0.photoslibrary/resources/derivatives/thumbs/4532.ithmb
    /Users/rhet/Pictures/Test-13.0.0.photoslibrary/resources/derivatives/masters/3
    /Users/rhet/Pictures/Test-13.0.0.photoslibrary/resources/derivatives/masters/3/3DD2C897-F19E-4CA6-8C22-B027D5A71907_4_5005_c.jpeg
    /Users/rhet/Pictures/Test-13.0.0.photoslibrary/resources/journals/HistoryToken.plist
    /Users/rhet/Pictures/Test-13.0.0.photoslibrary/resources/journals/Asset.plist
    /Users/rhet/Pictures/Test-13.0.0.photoslibrary/resources/journals/Asset-change.plj
    /Users/rhet/Pictures/Test-13.0.0.photoslibrary/private/com.apple.photolibraryd/caches/clientservertransaction
    /Users/rhet/Pictures/Test-13.0.0.photoslibrary/private/com.apple.photoanalysisd/caches/graph/PhotosGraph/photosgraph.kgdb-shm
```

In your case, only the AAE file or orientation change in the database is generated and Photos must be doing the render "just in time" when the user opens the photo in question.  When you use `--download-missing`, OSXPhotos will use AppleScript (unless you also use `--use-photokit` in which case the PhotoKit API is used) to instruct Photos to export the image. In your case this causes Photos to render the missing edited image and export it. In this situation `--download-missing` is a bit misleading because it's not downloading anything but it is forcing Photos to do the export for any missing assets which causes the render to happen.

So I think the solution for you is to simply continue using `--download-missing`. As for a more permanent solution for OSXPhotos, this is where I'm stuck on how to implement this. Monterey is an older OS and I no longer have access to a Monterey test machine. When I was first debugging this issue, I did have one and was not able to reproduce your observed behavior. Defaulting to `--download-missing` if on Monterrey and edited images are missing isn't elegant because it complicates the code but also because it will only work if the library being exported is the "last opened library" as that's the only library that can be scripted with AppleScript. If a user tried to export a non-default library, the export would fail because the UUID that OSXPhotos told Photos to export wouldn't be in the library that Photos is using.

I am (slowly) working on a new library over at [PhotoKit](https://github.com/RhetTbull/photokit) that would let me rewrite the core of OSXPhotos to use the same (undocumented) API that Photos uses to do all the exporting. This would be much more robust and would solve issues like this. It would also work with non-default libraries (that's required a fair bit of reverse engineering as all those APIs are undocumented).  However, it would also kill Linux support and support on older versions of macOS...so I might keep the "reaching into the database" version in the code as an alternate back-end. This is a slow effort as I don't have a lot of time to work on it and I'm stuck on a couple of key features that I'm still reverse engineering. Eventually I think I'll get it figured out though.

Bottom line: until I can create a more permanent fix, I suggest you use `--download-missing` on Monterey (as long as you're exporting the default library). Perhaps I should also add a note to the documentation about this in case others have the same issue.

**Comment by oPromessa** [comment]
Got it. Yes the renders either show up or not! That explains it. 

Thanks so much for the thorough feedback. 

Will happily use `--download-missing`

I'll eventually upgrade MacOS once I upgrade the hardware -- hopefully next year 😄

**Comment by oPromessa** [comment]
> I'll eventually upgrade MacOS once I upgrade the hardware -- hopefully next year 😄

So, I have finally upgraded the hardware and moved out of Monterey into macOS Tahoe 26.2. I no longer have missing "edited photo path".

> Here's an updated script to collect data on the adjustments. Please run this the same way and send the data. (This one won't output any paths).

Anyhow, decided to re-run `edited_paths.py` and it found 6 assets without an edited path field. 
Common thing is that they are all videos recorded by iPhones in **slomo**... not actually edited on Photos.
The slomo video is correctly exported. When you open them with QuickTime it shows some dots on the play bar which show which parts of the video are in slo-mo.

Their associated .plist file is something like:

```py
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>adjustmentBaseVersion</key>
	<integer>0</integer>
	<key>adjustmentData</key>
	<data>
(...)
	</data>
	<key>adjustmentEditorBundleID</key>
	<string></string>
	<key>adjustmentFormatIdentifier</key>
	<string>com.apple.video.slomo</string>
	<key>adjustmentFormatVersion</key>
	<string>1.1</string>
	<key>adjustmentRenderTypes</key>
	<integer>24576</integer>
	<key>adjustmentTimestamp</key>
	<date>2025-10-22T03:53:15Z</date>
</dict>
</plist>
```

Just wanted to let you know! And we can then close this issue.
